app/usecase/keys/consumer_cached.go

```diff
@@ -0,0 +1,81 @@
+package keys
+
+import (
+	"errors"
+)
+
+var _ Consumer = (*ConsumerCached)(nil)
+
+type bufferEntry struct {
+	key string
+	err error
+}
+
+type ConsumerCached struct {
+	delegate   Consumer
+	bufferSize int
+	buffer     chan bufferEntry
+}
+
+func (p ConsumerCached) ConsumeInBatch(maxCount uint) ([]string, error) {
+	keys := make([]string, 0, maxCount)
+
+	for ; maxCount > 0; maxCount-- {
+		// there is probability to get the list of keys with the size less than bufferSize
+		// in this case we listen to done channel to perform loop break
+		done := make(chan bool)
+
+		// we should load new keys as soon as our buffer is empty
+		if len(p.buffer) == 0 {
+			go func() {
+				p.loadKeys()
+				done <- true
+			}()
+		}
+
+		select {
+		case entry := <-p.buffer:
+			if entry.err != nil {
+				return keys, entry.err
+			}
+
+			keys = append(keys, entry.key)
+		case <-done:
+			break
+		}
```

Comment:
If  `maxCount` is two times of the bufferSize, will this work? Please wrap line 30 - 33 and line 36 - 44 with a function called refillKeys()

Reply:
Yes, it is going to work https://github.com/byliuyang/kgs/pull/41/files#diff-a13afc923323ecb98b71b95d485d6581R86 proves this statement.